### PR TITLE
Jobs Building Block & Scheduler Control Plane Service

### PR DESCRIPTION
**Pull request by cicoyle**
This PR adds the relevant code for:
1. a new Jobs Building Block 
2. a new Scheduler Control Plane Service

This PR follows [this official proposal,](https://github.com/dapr/proposals/pull/44). 

This is an end to end solution where an app is able to schedule a job to the daprd sidecar and perform basic crud operations with the job via the `Jobs api`. The daprd sidecar sends the job to the `Scheduler` control plane service, which then stores the job in its embedded etcd. At trigger time, the scheduler will send the job back to the daprd sidecar where the sidecar will send the triggered job along to the app. 

This includes work to replace the actor reminder subsystem, by scheduling the actor reminders in the Scheduler control plane service embedded etcd via the `SchedulerReminders` feature flag.

Perf numbers can be found in the [proposal comments here](https://github.com/dapr/proposals/pull/44).

There were several individuals involved in making this work happen - thank you to all 🚀 

## Issue reference

[1.14 release P0](https://github.com/dapr/dapr/issues/7605)
[Here is the issue tracking the work](https://github.com/dapr/dapr/issues/7246). There are a few remaining items that will be completed before the release date.

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: dapr/proto/runtime/v1/dapr.proto [hunk 1]
```diff
@@ -193,6 +193,15 @@ service Dapr {
...
   rpc Shutdown (ShutdownRequest) returns (google.protobuf.Empty) {}
+
+  // Create and schedule a job
+  rpc ScheduleJob(ScheduleJobRequest) returns (google.protobuf.Empty) {}
+
+  // Gets a scheduled job
+  rpc GetJob(GetJobRequest) returns (GetJobResponse) {}
+
+  // Delete a job
+  rpc DeleteJob(DeleteJobRequest) returns (google.protobuf.Empty) {}
```
Comment (mikeee) [inline dapr/proto/runtime/v1/dapr.proto:204]: Do these definitions need to be appended with `Alpha1`  as a suffix?

File: tests/perf/actor_reminder/actor_reminder_test.go [hunk 2]
```diff
@@ -152,3 +191,248 @@ func TestActorReminderRegistrationPerformance(t *testing.T) {
...
+		sidecarUsage,
+		appUsage)
+
+	report.SetTotalRestartCount(restarts)
+	err = utils.UploadAzureBlob(report)
+
+	if err != nil {
+		t.Error(err)
+	}
+
```
Comment (mikeee) [inline tests/perf/actor_reminder/actor_reminder_test.go:270]:
```suggestion
```

This blob upload step should be removed

File: charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  maxUnavailable: {{ .Values.global.ha.disruption.maximumUnavailable }}
+{{- end }}
+  selector:
+    matchLabels:
+      app: dapr-scheduler-server
+      {{- range $key, $value := .Values.global.k8sLabels }}
+      {{ $key }}: {{ tpl $value $ }}
+      {{- end }}
+      {{- with .Values.global.labels }}
+        {{- toYaml . | nindent 6 }}
```
Comment (ItalyPaleAle) [inline charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml:30]: Shouldn't this be 8?

File: cmd/scheduler/options/options.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	EtcdCompactionRetention string
+
+	Logger  logger.Options
+	Metrics *metrics.Options
+
+	taFile string
+}
+
+func New(origArgs []string) (*Options, error) {
+	// We are using pflag to parse the CLI flags
```
Comment (ItalyPaleAle) [inline cmd/scheduler/options/options.go:61]: This is probably not needed for a new service, since it was only for backwards-compat for daprd.

File: cmd/daprd/options/options.go [hunk 5]
```diff
@@ -248,6 +251,13 @@ func New(origArgs []string) (*Options, error) {
 		opts.DaprBlockShutdownDuration = nil
 	}
 
+	if !fs.Changed("scheduler-host-address") {
```
Comment (ItalyPaleAle) [inline cmd/daprd/options/options.go:254]: Since this uses StringSliceVar, the parsing of CSV values is automatic. StringSliceVar supports both CSV and repeating the flag

File: cmd/scheduler/options/options.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+	// Create a flag set
+	fs := pflag.NewFlagSet("scheduler", pflag.ExitOnError)
+	fs.SortFlags = true
+
+	fs.IntVar(&opts.Port, "port", 50006, "The port for the scheduler server to listen on")
+	fs.IntVar(&opts.HealthzPort, "healthz-port", 8080, "The port for the healthz server to listen on")
+	fs.StringVar(&opts.HealthzListenAddress, "healthz-listen-address", "", "The listening address for the healthz server")
+
+	fs.StringVar(&opts.ListenAddress, "listen-address", "127.0.0.1", "The address for the Scheduler to listen on")
```
Comment (ItalyPaleAle) [inline cmd/scheduler/options/options.go:86]: `0.0.0.0`? 127.0.0.1 is not good in containers (aside from sidecars)

File: dapr/proto/scheduler/v1/scheduler.proto [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  optional uint32 repeats = 2;
+
+  // Optional: sets time at which or time interval before the callback is invoked for the first time.
+  optional string due_time = 3;
+
+  // Optional: Time To Live to allow for auto deletes (accounting for Actor Reminders).
+  optional string ttl = 4;
+
+  // Job data.
+  google.protobuf.Any data = 5;
```
Comment (ItalyPaleAle) [inline dapr/proto/scheduler/v1/scheduler.proto:35]: Why this instead of `bytes`?

File: dapr/proto/scheduler/v1/scheduler.proto [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+message Job {
+  // The schedule for the job.
+  optional string schedule = 1;
+
+  // Optional: jobs with fixed repeat counts (accounting for Actor Reminders).
+  optional uint32 repeats = 2;
+
+  // Optional: sets time at which or time interval before the callback is invoked for the first time.
+  optional string due_time = 3;
```
Comment (ItalyPaleAle) [inline dapr/proto/scheduler/v1/scheduler.proto:29]: Should probably be a google.protobuf.Timestamp?

File: dapr/proto/scheduler/v1/scheduler.proto [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+  // actorTypes is the optional list of actor types to watch for.
+  repeated string actorTypes = 3;
+}
+
+// WatchJobsRequestResult is the result of a job execution to allow the job to
+// be marked as processed.
+message WatchJobsRequestResult {
+  // uuid is the uuid of the job that has finished processing.
+  uint64 uuid = 1;
```
Comment (ItalyPaleAle) [inline dapr/proto/scheduler/v1/scheduler.proto:99]: uuid is 128 bits so can't fit in uint64?

File: dapr/proto/scheduler/v1/scheduler.proto [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+// WatchJobsRequestInitial is the initial request to start watching for jobs.
+message WatchJobsRequestInitial {
+  // app_id is the App ID of the requester.
+  string app_id = 1;
+
+  // namespace is the namespace of the requester.
+  string namespace = 2;
+
+  // actorTypes is the optional list of actor types to watch for.
+  repeated string actorTypes = 3;
```
Comment (ItalyPaleAle) [inline dapr/proto/scheduler/v1/scheduler.proto:92]: `actor_types` because in protobuf the convention is to use snake_case

File: pkg/actors/internal/reminders.go [hunk 11]
```diff
@@ -33,6 +33,7 @@ type LookupActorFn func(ctx context.Context, actorType string, actorID string) (
 type StateStoreProviderFn func() (string, TransactionalStateStore, error)
 
 // RemindersProvider is the interface for the object that provides reminders services.
+// Deprecated: this will be removed after we have migrated to Scheduler Service.
```
Comment (ItalyPaleAle) [inline pkg/actors/internal/reminders.go:36]: There are dependencies on this, should not be deprecated

File: pkg/actors/actors.go [hunk 12]
```diff
@@ -1143,11 +1182,84 @@ func (a *actorsRuntime) doExecuteReminderOrTimer(ctx context.Context, reminder *
...
+
+	return ptr.Of("@every " + duration.String()), repeats, nil
+}
+
 func (a *actorsRuntime) CreateReminder(ctx context.Context, req *CreateReminderRequest) error {
 	if !a.actorsConfig.Config.HostedActorTypes.IsActorTypeHosted(req.ActorType) {
 		return ErrReminderOpActorNotHosted
 	}
 
+	if a.schedulerClients != nil && a.schedulerReminders {
```
Comment (ItalyPaleAle) [inline pkg/actors/actors.go:1212]: Please move to a separate function for readability

File: pkg/actors/actors.go [hunk 13]
```diff
@@ -230,6 +235,25 @@ func newActorsWithClock(opts ActorsOpts, clock clock.WithTicker) (ActorRuntime,
 
 	a.timers.SetExecuteTimerFn(a.executeTimer)
 
+	if opts.SchedulerReminders {
+		if opts.Config.SchedulerClients == nil {
+			return nil, fmt.Errorf("scheduler reminders are enabled, but no Scheduler clients are available")
+		}
+		log.Debug("Using Scheduler service for reminders.")
+		// We want to delete "a.actorsReminders" once we move to Scheduler service.
+		a.actorsReminders = reminders.NoOpReminders() // disable old reminder system if using Scheduler for reminders
```
Comment (ItalyPaleAle) [inline pkg/actors/actors.go:244]: Why wasn't this implemented using the `Reminders` interface? Deepanshu and I created that specifically so it could be used by other implementations. We have one and the idea was that the scheduler would use it too (and that's why it can't be deprecated). That would help keeping the code maintainable.

File: pkg/actors/actors.go [hunk 14]
```diff
@@ -1175,6 +1287,27 @@ func (a *actorsRuntime) DeleteReminder(ctx context.Context, req *DeleteReminderR
 		return ErrReminderOpActorNotHosted
 	}
 
+	if a.schedulerClients != nil && a.schedulerReminders {
+		internalDeleteJobReq := &schedulerv1pb.DeleteJobRequest{
```
Comment (ItalyPaleAle) [inline pkg/actors/actors.go:1291]: Ditto, please migrate to a separate function.

But--this should have been implemented in a struct that implements the `Reminder` interface

File: pkg/actors/actors.go [hunk 15]
```diff
@@ -1183,6 +1316,55 @@ func (a *actorsRuntime) GetReminder(ctx context.Context, req *GetReminderRequest
 		return nil, ErrReminderOpActorNotHosted
 	}
 
+	if a.schedulerClients != nil && a.schedulerReminders {
+		internalGetJobReq := &schedulerv1pb.GetJobRequest{
```
Comment (ItalyPaleAle) [inline pkg/actors/actors.go:1320]: Ditto

File: pkg/api/errors/errors.go [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+		httpCode,
+		message,
+		legacyTag,
+	).
+		WithErrorInfo(reason, metadata).
+		Build()
+}
+
+func NotConfigured(name string, componentType string, metadata map[string]string, grpcCode codes.Code, httpCode int, legacyTag string, reason string) error {
+	message := fmt.Sprintf("%s %s is not configured", componentType, name)
```
Comment (ItalyPaleAle) [inline pkg/api/errors/errors.go:46]: fmt.Sprintf should be avoided for simle string concatenation

```suggestion
	message := componentType + " " + name + " is not configured"
```

File: pkg/scheduler/server/internal/pool.go [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+			conns:     make(map[uint64]*conn),
+		}
+
+		p.nsPool[req.GetNamespace()] = nsPool
+	}
+
+	ok = true
+	var uuid uint64
+	for ok {
+		uuid = rand.Uint64() //nolint:gosec
```
Comment (ItalyPaleAle) [inline pkg/scheduler/server/internal/pool.go:94]: Should listen to the linter :) math/rand doesn't produce good random numbers. Use crypto/rand instead.
(But this should probably use a true UUID which has ~128 bits of entropy instead of 64...)

File: dapr/proto/runtime/v1/dapr.proto [hunk 18]
```diff
@@ -193,6 +193,15 @@ service Dapr {
   rpc RaiseEventWorkflowBeta1 (RaiseEventWorkflowRequest) returns (google.protobuf.Empty) {}
   // Shutdown the sidecar
   rpc Shutdown (ShutdownRequest) returns (google.protobuf.Empty) {}
+
+  // Create and schedule a job
+  rpc ScheduleJobAlpha1(ScheduleJobRequest) returns (google.protobuf.Empty) {}
```
Comment (ItalyPaleAle) [inline dapr/proto/runtime/v1/dapr.proto:198]: 1. Dont' return Emtpy. Instead, create a new message `ScheduleJobAlpha1Response` and make it empty. Same below.
2. Don't forget `Alpha` in the Request type

File: dapr/proto/internals/v1/reminders.proto [hunk 19]
```diff
@@ -35,3 +35,10 @@ message Reminder {
 message Reminders {
   repeated Reminder reminders = 1;
 }
+
+// TimerFiredEvent is the event that is sent to the actor when a timer fires.
+message TimerFiredEvent {
+  google.protobuf.Timestamp fireAt = 1;
```
Comment (ItalyPaleAle) [inline dapr/proto/internals/v1/reminders.proto:41]: `fire_at`

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml (hunk 3 above)
Comment (cicoyle) [reply charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml:30]: based on sentry, looks like 6 is fine: https://github.com/dapr/dapr/blob/master/charts/dapr/charts/dapr_sentry/templates/dapr_sentry_poddisruptionbudget.yaml#L30

File: cmd/daprd/options/options.go (hunk 5 above)
Comment (cicoyle) [reply cmd/daprd/options/options.go:254]: This is needed for the e2e tests to not break on the n-1 and n-2 versions since `scheduler-host-address` doesn't exist in those versions.

File: dapr/proto/internals/v1/reminders.proto (hunk 19 above)
Comment (cicoyle) [reply dapr/proto/internals/v1/reminders.proto:41]: I will update the protos to ensure the additions use `_`

File: dapr/proto/runtime/v1/dapr.proto (hunk 18 above)
Comment (cicoyle) [reply dapr/proto/runtime/v1/dapr.proto:198]: I don't see this being done for other protos, I'd prefer consistency. For example:

```
  rpc SubtleGetKeyAlpha1(SubtleGetKeyRequest) returns (SubtleGetKeyResponse);
  // SubtleEncryptAlpha1 encrypts a small message using a key stored in the vault.
  rpc SubtleEncryptAlpha1(SubtleEncryptRequest) returns (SubtleEncryptResponse);
```

File: charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml (hunk 3 above)
Comment (ItalyPaleAle) [reply charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml:30]: Ok I misunderstood the code. The issue is that there are 8 spaces at the start of the line and it caused confusion.
```suggestion
      {{- toYaml . | nindent 6 }}
```

File: cmd/daprd/options/options.go (hunk 5 above)
Comment (ItalyPaleAle) [reply cmd/daprd/options/options.go:254]: Not following sorry. That should be irrelevant for the n-1/n-2 tests? If it's passed or not...

File: charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml (hunk 3 above)
Comment (cicoyle) [reply charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml:30]: It works as is, are you saying there is an issue with indentation? It is consistent with the other pod disruption budgets in dapr.

File: cmd/daprd/options/options.go (hunk 5 above)
Comment (cicoyle) [reply cmd/daprd/options/options.go:254]: It needs to include this line due to the n-1/n-2 e2e tests not knowing the scheduler flag.

File: charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml (hunk 3 above)
Comment (ItalyPaleAle) [reply charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml:30]: It works because there's the `-` at the start of `{{`, but it is _technically_ confusing (since the line has 8 spaces at the start instead of 6). So more like a nit now.

File: dapr/proto/runtime/v1/dapr.proto (hunk 18 above)
Comment (ItalyPaleAle) [reply dapr/proto/runtime/v1/dapr.proto:198]: Ok. At least don't use `google.protobuf.Empty` please! This also helps with the tracing code as it expects a type for each message.

File: dapr/proto/scheduler/v1/scheduler.proto (hunk 8 above)
Comment (cicoyle) [reply dapr/proto/scheduler/v1/scheduler.proto:29]: The type is for consistency with actor reminders due time, which is string.

File: pkg/actors/internal/reminders.go (hunk 11 above)
Comment (yaron2) [reply pkg/actors/internal/reminders.go:36]: Indeed

File: dapr/proto/runtime/v1/dapr.proto (hunk 18 above)
Comment (yaron2) [reply dapr/proto/runtime/v1/dapr.proto:198]: Agree on not using Empty

File: pkg/actors/actors.go (hunk 13 above)
Comment (cicoyle) [reply pkg/actors/actors.go:244]: Of all funcs in the reminders interface, only 3 relate to scheduler: create, get, delete. It didnt make sense to implement the interface. The abstraction assumes the sidecars are responsible to keeping reminders and triggering them, but distributed scheduler elevates the abstraction.

File: pkg/actors/actors.go (hunk 12 above)
Comment (cicoyle) [reply pkg/actors/actors.go:1212]: sure thing

File: pkg/actors/actors.go (hunk 14 above)
Comment (cicoyle) [reply pkg/actors/actors.go:1291]: sure thing I can mv to a separate func, but plz see my response above as to why we didn't implement the reminder interface

File: pkg/actors/actors.go (hunk 15 above)
Comment (cicoyle) [reply pkg/actors/actors.go:1320]: same reply as above

File: pkg/api/errors/errors.go (hunk 16 above)
Comment (cicoyle) [reply pkg/api/errors/errors.go:46]: sure

File: cmd/scheduler/options/options.go (hunk 4 above)
Comment (cicoyle) [reply cmd/scheduler/options/options.go:61]: The comment or the `OrigArgs`?

Comment (ItalyPaleAle) [reply cmd/scheduler/options/options.go:61]: The entire block where the args are transformed. That was only needed in daprd, not for control plane services

File: pkg/actors/actors.go (hunk 13 above)
Comment (ItalyPaleAle) [reply pkg/actors/actors.go:244]: Please still implement the interface. Methods that do nothing can be no-ops. But using the interface allows the code to be much more readable and especially maintainable, since there are different implementations.

File: dapr/proto/runtime/v1/dapr.proto (hunk 18 above)
Comment (ItalyPaleAle) [reply dapr/proto/runtime/v1/dapr.proto:198]: Once a type is defined, it then needs to be added here for tracing: https://github.com/dapr/dapr/blob/master/pkg/proto/runtime/v1/dapr_tracing.go (it's ok to leave implementation as TODO for now)

**Comment by daixiang0** [comment]
@cicoyle maybe we can make all UT pass then do perf test.

**Comment by cicoyle** [comment]
> @cicoyle maybe we can make all UT pass then do perf test.

The unit tests are failing due to non Scheduler specific things. Specifically, its failing due to the dependency update. [See the fix in this PR](https://github.com/dapr/dapr/pull/7806). Also, there is some general flakiness in master with unit tests with the `informer` that is not related to Scheduler changes. [Workflows perf tests passed here](https://github.com/dapr/dapr/pull/7768#issuecomment-2158702313), we are doing some fine tuning gathering data points still.

File: charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml (hunk 3 above)
Comment (cicoyle) [reply charts/dapr/charts/dapr_scheduler/templates/dapr_scheduler_poddisruptionbudget.yaml:30]: I appreciate the nit. For consistency sake, I will leave this as is because all the other pod disruption budgets in dapr (specifically the `operator`, `injector`, `sentry`, and `placement`) all have this exact config. To avoid confusion, maybe all of them should be updated with your suggestion. However, I would like to avoid updating all of them in the scheduler PR

File: cmd/scheduler/options/options.go (hunk 4 above)
Comment (cicoyle) [reply cmd/scheduler/options/options.go:61]: From what I can find, it looks like the other control plane services include this diff as well. [See the PR where you added this to all the other control plane services.](https://github.com/dapr/dapr/commit/ace5b645edc32c1f95f6e689d1f65a939c47153e#diff-b0e58d8b097ed95fbd37b293e3e629e39829972fc09efce2360a15348ea03914) I understand that since Scheduler is a new service its not necessary to maintain backwards compatability, but maybe for consistency sake across the services it should remain- thoughts?

File: cmd/scheduler/options/options.go (hunk 6 above)
Comment (cicoyle) [reply cmd/scheduler/options/options.go:86]: I'm hesitant to put `0.0.0.0` as a default value due to the security implications.

File: dapr/proto/runtime/v1/dapr.proto (hunk 18 above)
Comment (cicoyle) [reply dapr/proto/runtime/v1/dapr.proto:198]: I have the TODOs in the `dapr_tracing` file in this PR currently. I can update it to not use Empty.

File: cmd/scheduler/options/options.go (hunk 6 above)
Comment (ItalyPaleAle) [reply cmd/scheduler/options/options.go:86]: The default listen address for the other CP services is `0.0.0.0`. It won't work in Docker otherwise.

File: cmd/scheduler/options/options.go (hunk 4 above)
Comment (ItalyPaleAle) [reply cmd/scheduler/options/options.go:61]: I'd remove it. That block is only useful for legacy reasons

File: dapr/proto/runtime/v1/dapr.proto (hunk 18 above)
Comment (ItalyPaleAle) [reply dapr/proto/runtime/v1/dapr.proto:198]: Yes do not use Empty, because it makes things easier in the future. Ok to leave TODOs in the tracing file for now

File: dapr/proto/internals/v1/reminders.proto (hunk 19 above)
Comment (cicoyle) [reply dapr/proto/internals/v1/reminders.proto:41]: https://github.com/cicoyle/dapr/pull/71

File: dapr/proto/runtime/v1/dapr.proto (hunk 18 above)
Comment (cicoyle) [reply dapr/proto/runtime/v1/dapr.proto:198]: https://github.com/cicoyle/dapr/pull/71

File: pkg/actors/internal/reminders.go (hunk 11 above)
Comment (cicoyle) [reply pkg/actors/internal/reminders.go:36]: https://github.com/cicoyle/dapr/pull/71

File: cmd/scheduler/options/options.go (hunk 6 above)
Comment (cicoyle) [reply cmd/scheduler/options/options.go:86]: [please see this PR with that change](https://github.com/cicoyle/dapr/pull/71/files#diff-0de3b642b59f384f86664dea2875d7602cdc27323f3220286b52ce0126c758f8)

File: cmd/scheduler/options/options.go (hunk 4 above)
Comment (yaron2) [reply cmd/scheduler/options/options.go:61]: @cicoyle was this change request addressed?

File: cmd/scheduler/options/options.go (hunk 6 above)
Comment (cicoyle) [reply cmd/scheduler/options/options.go:71]: After further consideration, I left this as is since as a user I would not want anything listening on `0.0.0.0` by default bc if I forgot to change the default value then that is a security concern. Does anyone else feel strongly about this and could weigh in? cc/ @daixiang0 @yaron2 @artursouza @JoshVanL @mukundansundar

File: dapr/proto/scheduler/v1/scheduler.proto (hunk 9 above)
Comment (cicoyle) [reply dapr/proto/scheduler/v1/scheduler.proto:99]: This is used as an incremental counter, not really a `uuid`. I can change the name to be job `id` and put an explicit note in the comment description for it, or do you have a suggestion for a different name?

Comment (ItalyPaleAle) [reply dapr/proto/scheduler/v1/scheduler.proto:99]: Yes please, rename it. `id` or `seq` (sequence number) could work. The name `uuid` implies it's a UUID which is not a number

File: pkg/scheduler/server/internal/pool.go (hunk 17 above)
Comment (cicoyle) [reply pkg/scheduler/server/internal/pool.go:94]: This is used as indexing rather than a uuid. I can update the name to just be id to avoid confusion. Since there are no security implication here and it is used for indexing purposes, I can change this to id++.

Comment (ItalyPaleAle) [reply pkg/scheduler/server/internal/pool.go]: Ok `id++` is fine.

It's not just about security, it's also that `math/rand` doesn't guarantee good randomness, so conflicts can arise. If you can use an increment, that's better

File: cmd/scheduler/options/options.go (hunk 6 above)
Comment (artursouza) [reply cmd/scheduler/options/options.go:71]: Alessandro is correct that we need to listen to all addresses for this control plane to run from the container without having to change CLI to pass additional params (it would be a breaking change too for CLI to work with older Dapr versions or old CLI to work with newer Dapr versions).

All other control plane services (right now) listen on "" for this reason. I suggest we keep scheduler consistent (in this PR) and also listen on "". Then we have a separate PR to change all control planes (and the shared Dockerfile) to have them to listen to "" automatically from inside the docker image, without CLI having to pass additional params. We can have a new env var named DAPR_CONTROL_PLANE_LISTEN_ADDR or something that only control plane services will understand and it is set in Dockerfile globally.

Comment (JoshVanL) [reply cmd/scheduler/options/options.go:71]: Software generally should not be made to listen on all interfaces and therefore be publicly available on the network by default. This is a clear security issue. Dapr services which are supported to be run from the CLI should default `localhost` for listening addresses. In the case of both kube and container target deployments, we have control over what arguments are given to the process via Helm and `dapr` CLI.

We should change `sentry`, `placement`, `daprd` and `scheduler` listen address defaults to `localhost`. I don't think there is a need for more config surface to be added because CLI flags already exist to change these values.

Comment (cicoyle) [reply cmd/scheduler/options/options.go]: I've updated the default for Scheduler to be `""` for consistency. Although, we should revisit changing the defaults separate of the Scheduler PR

File: tests/integration/integration.go [hunk 20]
```diff
@@ -74,7 +74,7 @@ func RunIntegrationTests(t *testing.T) {
 			options := tcase.Setup(t)
 
 			t.Logf("setting up test case")
-			ctx, cancel := context.WithTimeout(context.Background(), 45*time.Second)
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
```
Comment (JoshVanL) [inline tests/integration/integration.go:77]: Please revert this change. No integration test should be getting close to 45s, and is already a generous upper bound. This is here so that buggy code/tests don't make runs go on forever.

**Review by JoshVanL — CHANGES_REQUESTED** [review]

File: cmd/scheduler/options/options.go (hunk 6 above)
Comment (ItalyPaleAle) [reply cmd/scheduler/options/options.go]: > Software generally should not be made to listen on all interfaces and therefore be publicly available on the network by default. 

While this is true for "regular" binaries, the difference is that Dapr control plane components are only meant to run inside Docker containers, even in self-hosted mode. We do not support running outside of a Docker container. Inside Docker, binding to 127.0.0.1 is not the desired behavior in almost all cases; if users want to make sure the container isn't exposed to the world, that should be done on Docker and not the binary.

Comment (artursouza) [reply cmd/scheduler/options/options.go]: > Software generally should not be made to listen on all interfaces and therefore be publicly available on the network by default. This is a clear security issue. Dapr services which are supported to be run from the CLI should default `localhost` for listening addresses. In the case of both kube and container target deployments, we have control over what arguments are given to the process via Helm and `dapr` CLI.
> 
> We should change `sentry`, `placement`, `daprd` and `scheduler` listen address defaults to `localhost`. I don't think there is a need for more config surface to be added because CLI flags already exist to change these values.

Like I mentioned, requiring new params from CLI would be a breaking change. The CLI is meant to work with Dapr runtime version +-1, the container image should be done self-sufficient to run all binaries without extra params. I would not be in favor of breaking this compatibility if the cost is just a new env that users won't even need to set and come pre-set in the container image.

Anyway, for the purpose of this change, consistency is better than fixing for only one service and having to write on-off logic for scheduler.

Comment (artursouza) [reply cmd/scheduler/options/options.go]: > > Software generally should not be made to listen on all interfaces and therefore be publicly available on the network by default.
> 
> While this is true for "regular" binaries, the difference is that Dapr control plane components are only meant to run inside Docker containers, even in self-hosted mode. We do not support running outside of a Docker container. Inside Docker, binding to 127.0.0.1 is not the desired behavior in almost all cases; if users want to make sure the container isn't exposed to the world, that should be done on Docker and not the binary.

Users can download the binaries and run it themselves in standalone mode without using Docker at all, but I agree that this would be rare and have not heard about it yet. Just a possible scenario.

Comment (artursouza) [reply cmd/scheduler/options/options.go]: Moving the discussion on "if" and "how" to change control planes to listen to localhost to this issue: https://github.com/dapr/dapr/issues/7834

For the sake of this PR, can we all agree to keep consistency? 👍 or 👎

Comment (yaron2) [reply cmd/scheduler/options/options.go]: I agree that it's a much better security posture to change to localhost across all control plane services, and that it should be done, outside the scope of this PR.

File: dapr/proto/runtime/v1/appcallback.proto [hunk 21]
```diff
@@ -45,6 +46,9 @@ service AppCallback {
   // User application can save the states or send the events to the output
   // bindings optionally by returning BindingEventResponse.
   rpc OnBindingEvent(BindingEventRequest) returns (BindingEventResponse) {}
+
+  // Sends job back to the app's endpoint at trigger time.
+  rpc OnJobEvent (JobEventRequest) returns (JobEventResponse);
```
Comment (artursouza) [inline dapr/proto/runtime/v1/appcallback.proto:51]: This needs to be moved to AppCallbackAlpha since it is as Alpha API

**Review by artursouza — COMMENTED** [review]
Only this comment. Let me try to make this change and merge it now.

File: dapr/proto/runtime/v1/appcallback.proto (hunk 21 above)
Comment (artursouza) [reply dapr/proto/runtime/v1/appcallback.proto]: Just made the changes. Observing the build.

**Comment by cicoyle** [comment]
Looks like the integration tests that are failing are unrelated: 
`Test_Integration/placement/vnodes/vnodes` && 
`Test_Integration/daprd/secret/http/componentName` && `Test_Integration/daprd/hotreload/operator/informer/scopes/components`

The e2e test `TestPubSubGRPC` failure is unrelated as well I believe

**Comment by yaron2** [comment]
@cicoyle see minor conflict needs resolving

**Comment by cicoyle** [comment]
looks like that int test is flaky and unrelated: `tests/integration.Test_Integration/placement/dissemination/notls`
